Kubelet: implement support for podAndContainerStatsFromCRI

haircommander opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature
/kind design

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->
/kind feature
#### What this PR does / why we need it:
This PR uses the CRI changes in https://github.com/kubernetes/kubernetes/pull/102789 to implement the feature gate for kubernetes/enhancements#2364  
#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Add support for PodAndContainerStatsFromCRI featuregate, which allows a user to specify their pod stats must also come from the CRI, not cAdvisor.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs


- [KEP]: https://github.com/kubernetes/enhancements/issues/2371
```

haircommander committed:
features: add podAndContainerStatsFromCRI

to allow users to specify whether the Kubelet should pull pod and container stats
strictly from the CRI, rather than a mixture of CRI and cAdvsior

Signed-off-by: Peter Hunt <pehunt@redhat.com>

haircommander committed:
kubelet stats: wire up podAndContainerStatsFromCRI feature gate

though it is currently unused

Signed-off-by: Peter Hunt <pehunt@redhat.com>

haircommander committed:
kubelet stats: add support for podAndContainerStatsFromCRI

This commit adds an initial implementation of translating from the new CRI fields
to the /stats/summary PodStats object

Signed-off-by: Peter Hunt <pehunt@redhat.com>

haircommander commented:
/sig node
/priority important-soon

haircommander commented on staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto:
```diff
@@ -1272,6 +1385,15 @@ message FilesystemUsage {
     // This may not equal InodesCapacity - InodesAvailable because the underlying
     // filesystem may also be used for purposes other than storing images.
     UInt64Value inodes_used = 4;
+    // TODO: Unclear how the remaining fields relate to container stats. Is it filled in cAdvisor?
```
also TODO: I have yet to actually figure out how to wire these up yet. it's not clear this should be the responsibility of CRI impl or cAdvisor

mrunalp replied:
Does this make more sense as a cumulative stat rather than associating with an ImageSpec for a container?

ehashman commented:
/triage accepted

haircommander committed:
kubelet: slightly move around some cri stats functions

to reduce duplication and add clarity

Signed-off-by: Peter Hunt <pehunt@redhat.com>

haircommander committed:
kubelet stats: use UsageNanoCores if available

Signed-off-by: Peter Hunt <pehunt@redhat.com>

haircommander commented:
/retest

haircommander committed:
kubelet: use helper function to check for nil fields in sandbox stats

Signed-off-by: Peter Hunt <pehunt@redhat.com>

haircommander commented:
/retest

mrunalp commented on staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto:
```diff
@@ -545,6 +551,114 @@ message ListPodSandboxResponse {
+    repeated PodSandboxStats stats = 1;
+}
+
+// PodSandboxAttributes provides basic information of the container.
```
nit: `of the pod`

mrunalp commented on staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto:
```diff
@@ -545,6 +551,114 @@ message ListPodSandboxResponse {
+    CpuUsage cpu = 1;
+    // Memory usage gathered from the pod.
+    MemoryUsage memory = 2;
+    // Stats pertaining to CPU resources consumed by pod cgroup (which includes all containers' resource usage and pod overhead).
```
This comment is for `MemoryUsage`. Can you move it up and add right one for network stats?

mrunalp commented on staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto:
```diff
@@ -545,6 +551,114 @@ message ListPodSandboxResponse {
+    // The time at which these stats were updated.
+    int64 timestamp = 1;
+
+    // Stats for the default interface, if found
```
Is `if found` needed? (There will always be a default interface, right?)

mrunalp commented on staging/src/k8s.io/cri-api/pkg/apis/runtime/v1alpha2/api.proto:
```diff
@@ -549,6 +555,114 @@ message ListPodSandboxResponse {
+    repeated InterfaceStats interfaces = 3;
+}
+
+// InterfaceStats contains resource value data about interface.
```
nit: `about a network interface`

mrunalp commented on staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto:
```diff
@@ -1392,15 +1392,6 @@ message FilesystemUsage {
     // This may not equal InodesCapacity - InodesAvailable because the underlying
     // filesystem may also be used for purposes other than storing images.
     UInt64Value inodes_used = 4;
```
Can we fold this commit into the first one?

mrunalp commented on staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto:
```diff
@@ -614,9 +614,9 @@ message LinuxPodSandboxStats {
     MemoryUsage memory = 2;
     // Stats pertaining to CPU resources consumed by pod cgroup (which includes all containers' resource usage and pod overhead).
-    NetworkStats network = 3;
+    NetworkUsage network = 3;
```
Can we merge this commit into the first commit?

bobbypage commented on pkg/kubelet/cri/remote/remote_runtime.go:
```diff
@@ -569,6 +569,46 @@ func (r *remoteRuntimeService) ListContainerStats(filter *runtimeapi.ContainerSt
+	return resp.GetStats(), nil
+}
+
+// ListContainerStats returns the list of ContainerStats given the filter.
```
godoc comment is out of date

```
ListPodSandboxStats returns the list of pod sandbox stats given the filter
```

bobbypage commented on pkg/kubelet/stats/cri_stats_provider.go:
```diff
@@ -830,3 +889,71 @@ func getCRICadvisorStats(infos map[string]cadvisorapiv2.ContainerInfo) map[strin
+		UsageCoreNanoSeconds: valueOfUInt64Value(criCPU.UsageCoreNanoSeconds),
+	}
+
+	if criPodStat == nil || criPodStat.Linux == nil || criPodStat.Linux.Memory == nil {
```
we already have the `criPodStat == nil || criPodStat.Linux == nil` checks above on line 920, i think this is redundant

haircommander replied:
however, now that they're separated, they are both now needed

bobbypage commented on pkg/kubelet/stats/cri_stats_provider.go:
```diff
@@ -830,3 +889,71 @@ func getCRICadvisorStats(infos map[string]cadvisorapiv2.ContainerInfo) map[strin
+	}
+}
+
+func addCRIPodCPUAndMemoryStats(ps *statsapi.PodStats, criPodStat *runtimeapi.PodSandboxStats) {
```
nit: i would separate this two functions, one for cpu stats, one for memory stats

haircommander replied:
yeah i find it weird the cadvisor variants of this don't do that.

bobbypage commented on pkg/kubelet/stats/cri_stats_provider.go:
```diff
@@ -634,6 +686,10 @@ func (p *criStatsProvider) getContainerUsageNanoCores(stats *runtimeapi.Containe
 		return nil
 	}
 
+	if stats.Cpu != nil && stats.Cpu.UsageNanoCores != nil {
```
just to understand, reason we don't need to acquire mutex etc, is because we are not using the "cached" `usageNanoCores` and instead using one directly supplied from the stats message?

bobbypage replied:
is there any reason for us to cache this `UsageNanoCores` in CRI stats world or is it unnecessary now?

haircommander replied:
I don't see much need to deal with locking if we're going to bypass the cache now

bobbypage commented on pkg/kubelet/stats/cri_stats_provider.go:
```diff
@@ -212,45 +204,80 @@ func (p *criStatsProvider) listPodStats(updateCPUNanoCoreUsage bool) ([]statsapi
+func (p *criStatsProvider) ListPodCPUAndMemoryStats() ([]statsapi.PodStats, error) {
 	// sandboxIDToPodStats is a temporary map from sandbox ID to its pod stats.
 	sandboxIDToPodStats := make(map[string]*statsapi.PodStats)
+	containerMap, podSandboxMap, err := p.getPodAndContainerMaps()
```
just to confirm the reason that the is removed from the diff is because this logic is now handled by `getPodAndContainerMaps`, correct?

```
containers = removeTerminatedContainers(containers)
	// Creates container map.
	containerMap := make(map[string]*runtimeapi.Container)
	for _, c := range containers {
		containerMap[c.Id] = c
	}
```

haircommander replied:
yup!

bobbypage commented on pkg/kubelet/stats/cri_stats_provider.go:
```diff
@@ -295,13 +322,38 @@ func (p *criStatsProvider) ListPodCPUAndMemoryStats() ([]statsapi.PodStats, erro
+		return nil, nil, fmt.Errorf("failed to list all containers: %v", err)
+	}
+
+	// Creates pod sandbox map.
```
can we note what is key -> value here?

haircommander replied:
added

bobbypage commented on pkg/kubelet/stats/cri_stats_provider.go:
```diff
@@ -295,13 +322,38 @@ func (p *criStatsProvider) ListPodCPUAndMemoryStats() ([]statsapi.PodStats, erro
+	}
+
+	containers = removeTerminatedContainers(containers)
+	// Creates container map.
```
ditto here, note what is what is key -> value mapping

bobbypage commented on pkg/features/kube_features.go:
```diff
@@ -761,6 +761,13 @@ const (
+	// kep: http://kep.k8s.io/2364
+	// alpha: v1.22
+	//
+	// Configures the Kubelet to use the CRI to populate pod and container stats, instead of supplimenting with stats from cAdvisor.
```
Maybe note that this requires CRI runtime to support pod and container level stats.

bobbypage commented on pkg/kubelet/dockershim/docker_stats.go:
```diff
@@ -62,3 +65,15 @@ func (ds *dockerService) ListContainerStats(ctx context.Context, r *runtimeapi.L
+	return nil, ErrNotImplemented
+}
+
+// ListPodSandboxStats returns stats for a list container stats request based on a filter.
```
ListPodSandboxStats returns stats for a list pod sandbox stats request...

bobbypage commented on pkg/kubelet/stats/cri_stats_provider.go:
```diff
@@ -295,13 +322,38 @@ func (p *criStatsProvider) ListPodCPUAndMemoryStats() ([]statsapi.PodStats, erro
+		return nil, nil, fmt.Errorf("failed to list all containers: %v", err)
+	}
+
+	// Creates pod sandbox map.
```
ditto, add doco on key -> value mapping of this map and below one

haircommander replied:
added

bobbypage commented on pkg/kubelet/stats/cri_stats_provider.go:
```diff
@@ -634,6 +686,10 @@ func (p *criStatsProvider) getContainerUsageNanoCores(stats *runtimeapi.Containe
 		return nil
 	}
 
+	if stats.Cpu != nil && stats.Cpu.UsageNanoCores != nil {
```
add comment to explain why there's two code paths here and why in this case we avoid using cached `UsageNanoCores`

bobbypage commented on pkg/kubelet/stats/cri_stats_provider.go:
```diff
@@ -653,6 +709,9 @@ func (p *criStatsProvider) getAndUpdateContainerUsageNanoCores(stats *runtimeapi
 	if stats == nil || stats.Attributes == nil || stats.Cpu == nil || stats.Cpu.UsageCoreNanoSeconds == nil {
 		return nil
 	}
+	if stats.Cpu.UsageNanoCores != nil {
```
ditto here, worth to explain that cache is unused

haircommander replied:
done!

bobbypage commented on staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto:
```diff
@@ -545,6 +551,113 @@ message ListPodSandboxResponse {
     repeated PodSandbox items = 1;
 }
 
+message PodSandboxStatsRequest {
```
I assume proto changes will not be included in this PR but rather in #102789 instead correct?

haircommander replied:
yeah this is based on that

haircommander commented:
/retest

haircommander commented:
/retest

mikebrow reviewed (approved):
LGTM
Noting: Per the KEP.. this alpha stage has the feature gate (PodAndContainerStatsFromCRI) off by default so we can discover/work on performance issues and expand testing for when the feature gate gets enabled by default in a later release.

haircommander commented:
/retest

haircommander commented:
implements many of the alpha pieces of  https://github.com/kubernetes/enhancements/issues/2371

haircommander committed:
kubelet: fallback to partial CRI stats if full fails

This is partially to allow the kube alpha tests to pass until CRI implementations have support, but also to handle this error situation a bit more elegantly

Signed-off-by: Peter Hunt <pehunt@redhat.com>

rphillips commented:
This looks ready.

/lgtm

bobbypage commented on pkg/kubelet/stats/cri_stats_provider.go:
```diff
@@ -122,21 +127,29 @@ func (p *criStatsProvider) listPodStats(updateCPUNanoCoreUsage bool) ([]statsapi
+			}
+			// CRI implementation doesn't support ListPodSandboxStats, warn and fallback.
+			klog.V(5).ErrorS(err,
+				"CRI implementation must be updated to support ListPodSandboxStats if PodAndContainerStatsFromCRI feature gate is enabled. Falling back to populating with cAdvisor; this call will fail in the future.",
```
+1 on adding this as a fallback

bobbypage commented:
LGTM as well, thanks @haircommander for your work on this. 

We plan to add new E2E tests that will validate CRI stats using running against updated container runtime implementations that support the new endpoints.

/lgtm

derekwaynecarr commented:
thanks @haircommander and @bobbypage for the mutual work on this.
/approve
/lgtm
